src/transaction/log_replication_atomic.cpp

```diff
@@ -126,6 +127,10 @@ namespace cublog
 	    if (m_locked_classes.count (header.trid) > 0)
 	      {
 		discard_caches_for_ddl (thread_entry, header.trid);
+	      }
```

Comment:
Nit.

What is the point of these checks done here? Can't the checks be moved to within the function. Nothing against, but I don't see the purpose of this verbosity and I am curious.

Reply:
DDL has a lower execution rate than DML, and I thought that calling DDL functions for every commit/abort log would incur costs. There are stack creations/destructions, changes in register values, etc. If the caller knows the conditions for calling the function, I wanted to selectively call the function based on these conditions. If the specified conditions are not suitable for the context, are not understandable, or are too verbose, would it be better to place them inside?

Reply:
Your logic makes sense. I would have done the same to optimize a bit.

But then, it makes more sense to reuse the result of `m_locked_classes.count (header.trid)` instead of calling it repeatedly (called at ~127 and then again at ~132).
The compiler might be able to optimize that in release compilation, but I am not sure since there is some side effect inbetween.

Better still, in `atomic_replicator::discard_caches_for_ddl` an equal range is extracted for `trid` that can also be used to infer the result of `count (header.trid)` - `count()` value is not actually needed, what is needed is "is there at least one element in the range".

In general whatever you do in a function as processing far outweighs the cost of calling that function.
Compare current implementation, for each LOG_COMMIT/LOG_ABORT:
- call to `m_locked_classes.count`
- a second call to `m_locked_classes.count` (which could be avoided) and a call to `m_locked_serials.count`

versus a possible:
- call to `atomic_replicator::discard_caches_for_ddl`
- call to `equal_range`, find if the range contains anything by iterating and breaking it (as far as I know, there is no `std::...` algorithm to answer the question "do I have at least one element in this range of iterators"
- based o that, call, or not `discard_caches_for_ddl` with the equal range you already have
- and something similar for `m_locked_serials` ..
- I think that, all in all would result in a more optimized execution, but it's just nit-picking :)

The cheapest thing is to reuse the result of `m_locked_classes.count` here and - probably - for abort as well.

Reply:
> versus a possible:

call to atomic_replicator::discard_caches_for_ddl
call to equal_range, find if the range contains anything by iterating and breaking it (as far as I know, there is no std::... algorithm to answer the question "do I have at least one element in this range of iterators"
based o that, call, or not discard_caches_for_ddl with the equal range you already have
and something similar for m_locked_serials ..
I think that, all in all would result in a more optimized execution, but it's just nit-picking :)

I agree with you, there seem to be areas where optimization can be applied. 
For now, I will reuse the `m_locked_classes.count`